### PR TITLE
e2etest: implement testTokenTransactions from vochaintest

### PR DESCRIPTION
**Pull request by altergui**
this also includes a small refactor to allow more tests to be added in the future

### CODE REVIEW

**Comment by altergui** [comment]
i'm soooo confident this works, i even added to testsuite so it's run in CI :innocent:

File: apiclient/blockchain.go [hunk 1]
```diff
@@ -8,13 +8,89 @@ import (
...
+	case models.TxType_NEW_PROCESS:
+		return txscost.Costs.NewProcess, nil
+	case models.TxType_ADD_DELEGATE_FOR_ACCOUNT:
+		return txscost.Costs.AddDelegateForAccount, nil
+	case models.TxType_DEL_DELEGATE_FOR_ACCOUNT:
+		return txscost.Costs.DelDelegateForAccount, nil
+	case models.TxType_COLLECT_FAUCET:
+		return txscost.Costs.CollectFaucet, nil
+	default:
+		return 0, fmt.Errorf("transaction type unknown")
```
Comment (altergui) [inline apiclient/blockchain.go]: this is sooo horrible. any ideas?

Comment (p4u) [reply apiclient/blockchain.go]: see vochain/state/balances.go

File: cmd/end2endtest/account.go [hunk 2]
```diff
@@ -1,32 +1,277 @@
...
+	}
+}
+
+func testGetTxCost(api *apiclient.HTTPclient) error {
+	// get treasurer nonce
+	treasurer, err := api.Treasurer()
+	if err != nil {
+		return err
+	}
+	log.Infof("treasurer is %s", common.BytesToAddress(treasurer.Address))
```
Comment (p4u) [inline cmd/end2endtest/account.go]: from now on, we are trying to use the key/value log functions when printing information with values, such as `log.Infow("treasurer", map[string]interface{}{"address": common.BytesToAddress(treasurer.Address)}`

Also use `log.Errorw(err, "message")` for printing errors.

File: api/api_types.go [hunk 3]
```diff
@@ -158,6 +158,23 @@ type TransactionReference struct {
 	Height uint32 `json:"blockHeight"`
 	Index  uint32 `json:"transactionIndex"`
 }
+type TransactionsCost struct {
+	Costs transactionCosts `json:"costs,omitempty"`
+}
+type transactionCosts struct {
```
Comment (p4u) [inline api/api_types.go]: Is this definition actually needed? Cannot we manage it with a `map[string]uint64` api/chain.go does?

File: apiclient/blockchain.go [hunk 4]
```diff
@@ -8,13 +8,89 @@ import (
...
+	}
+}
+
+// SubmitTx POSTs the given tx to the API, and returns txHash and response
+func (c *HTTPclient) SubmitTx(stx *models.SignedTx) (types.HexBytes, []byte, error) {
```
Comment (p4u) [inline apiclient/blockchain.go]: Is this method needed?

Usually in the apiclient we try to abstract the user from dealing with transactions. So if there is a need for sending a signed transaction, a specific method should be implemented. 

For dealing with multiple keys to perform different signatures you can use: `c.Clone("newPrivKey").Whatever()`

File: cmd/end2endtest/account.go [hunk 5]
```diff
@@ -1,32 +1,277 @@
...
+	if err := bob.Generate(); err != nil {
+		log.Fatal(err)
+	}
+
+	// Connect to the API host
+	hostURL, err := url.Parse(c.host)
+	if err != nil {
+		log.Fatal(err)
+	}
+	log.Debugf("connecting to %s", hostURL.String())
```
Comment (p4u) [inline cmd/end2endtest/account.go]: log.Debugw() will look nice here :)

**Review by p4u — APPROVED** [review]

**Comment by altergui** [comment]
closed due to branch rename, followup in #793

File: apiclient/blockchain.go (hunk 4 above)
Comment (altergui) [reply apiclient/blockchain.go]: definitely not needed, actually unused. dropped it

File: api/api_types.go (hunk 3 above)
Comment (altergui) [reply api/api_types.go]: loved this solution. made everything so simpler, thanks a million

File: cmd/end2endtest/account.go (hunk 2 above)
Comment (altergui) [reply cmd/end2endtest/account.go]: right, forgot about that. will do :+1:

**Pull request closed without merging**